### ISSUE TITLE
Nested Lists are wrongly derived as Nil, head and tl$1

### ISSUE THREAD

**Issue by fernandomora** [circe/circe#444]
This is tested in circe 0.6.0-RC1 with scala 2.11

```
scala> import io.circe._, io.circe.generic.auto._, io.circe.parser._, io.circe.syntax._
import io.circe._
import io.circe.generic.auto._
import io.circe.parser._
import io.circe.syntax._

scala> case class Test(list: List[Test])
defined class Test

scala> List(Test(List.empty[Test])).asJson
res0: io.circe.Json =
[
  {
    "list" : {
      "Nil" : {

      }
    }
  }
]

scala> decode[List[Test]]("""[{"list":[]}]""")
res1: Either[io.circe.Error,List[Test]] = Left(DecodingFailure(CNil, List(El(DownField(list),true,false), El(DownArray,true,false))))

scala> decode[List[Test]]("""[{"list":{"Nil":{}}}]""")
res2: Either[io.circe.Error,List[Test]] = Right(List(Test(List())))
```

But `asJson` should produce:
```
[
  {
    "list" : [
    ]
  }
]
```

And `decode` should be able to read:
```
[
  { "list": [] }
]
```

Non-empty lists also are wrongly derived:
```
scala> List(Test(List(Test(Nil)))).asJson
res13: io.circe.Json =
[
  {
    "list" : {
      "::" : {
        "head" : {
          "list" : {
            "Nil" : {

            }
          }
        },
        "tl$1" : {
          "Nil" : {

          }
        }
      }
    }
  }
]
```

**Comment by fernandomora** [comment]
As a workaround if If you explicitly declare the codecs using `decodeCanBuildFrom` and it works as expected:
```
scala> import io.circe._, io.circe.generic.auto._, io.circe.parser._, io.circe.syntax._
import io.circe._
import io.circe.generic.auto._
import io.circe.parser._
import io.circe.syntax._

scala> case class Test(list: List[Test])
defined class Test

scala> implicit val decoder: Decoder[List[Test]] = Decoder.decodeCanBuildFrom[Test, List]
decoder: io.circe.Decoder[List[Test]] = io.circe.SeqDecoder@3f6566fe

scala> implicit val encoder: Encoder[List[Test]] = Encoder.encodeTraversableOnce[Test, List]
encoder: io.circe.Encoder[List[Test]] = io.circe.MidPriorityEncoders$$anon$10@46523512

scala> List(Test(List.empty[Test])).asJson
res0: io.circe.Json =
[
  {
    "list" : [
    ]
  }
]

scala> List(Test(List(Test(Nil)))).asJson
res1: io.circe.Json =
[
  {
    "list" : [
      {
        "list" : [
        ]
      }
    ]
  }
]

scala> decode[List[Test]]("""[{"list":[]}]""")
res2: Either[io.circe.Error,List[Test]] = Right(List(Test(List())))
```

**Comment by travisbrown** [comment]
I think this is the same underlying issue as in #390, but in a slightly more minimized form. Thanks for identifying the workaround, @fernandomora.

I don't think it'll be possible to fix this in 0.6.0, but we should try to characterize the cases where it occurs more specifically and write up workarounds.

**Comment by travisbrown** [comment]
Confirmed that this is resolved in all current Scala version on master.
